File: src/main/java/xyz/deseteral/lebkuchenfm/domain/CommandExecutor.java
package xyz.deseteral.lebkuchenfm.domain;

import org.springframework.stereotype.Component;

import java.util.List;

@Component
public class CommandExecutor {
    private final List<CommandProcessor> commandProcessors;

    public CommandExecutor(List<CommandProcessor> commandProcessors) {
        this.commandProcessors = List.copyOf(commandProcessors);
    }

    public CommandProcessingResponse process(Command command) {
        return commandProcessors.stream()
            .filter(processor -> processor.matches(command))
            .findFirst()
            .map(processor -> processor.process(command.getArgs()))
            .orElseThrow(() -> new NoSuchCommandProcessorException(command));
    }

    public CommandProcessingResponse processFromText(String commandText) {
        return CommandParser.parse(commandText)
            .map(this::process)
            .orElseThrow(() -> new TextIsNotACommandException(commandText));
    }
}

File: src/main/kotlin/xyz/deseteral/lebkuchenfm/domain/CommandExecutor.kt
package xyz.deseteral.lebkuchenfm.domain

import org.springframework.stereotype.Component

@Component
class CommandExecutor(commandProcessors: List<CommandProcessor>) {
    private val commandProcessors: List<CommandProcessor>

    init {
        this.commandProcessors = List.copyOf<CommandProcessor>(commandProcessors)
    }

    fun process(command: Command): CommandProcessingResponse {
        return commandProcessors.stream()
                .filter { processor -> command.matchProcessor(processor) }
                .findFirst()
                .map { processor -> processor.process(command.args) }
                .orElseThrow { NoSuchCommandProcessorException(command) }
    }

    fun processFromText(commandText: String): CommandProcessingResponse {
        return CommandParser.parse(commandText)
                .map<CommandProcessingResponse>(Function<Command, CommandProcessingResponse> { this.process(it) })
                .orElseThrow { TextIsNotACommandException(commandText) }
    }
}
